Add untested files from specific directories

wesen1 opened the pull request:
My projects have a structure like this:

* src
* tests
  * .luacov

I run luacov from the tests directory and use the `include` config option to include all files from `../src`.
I tried using the new `includeuntestedfiles` config option but it had no effect. So I checked the code and noticed that if the `includeuntestedfiles` option is enabled only the lua files in the current working directory will be added recursively to the result.

The changes of this Pull Request will make the `includeuntestedfiles` option add only the lua files from the `include` directories if the `include` option is set, otherwise the previous behaviour will be used.

Tested the changes by using `luarocks make` and running the tests of one of my projects.

hishamhm commented:
Hi, thanks for contributing! The change in the PR is welcome, but the code change does not follow the coding style of the project:

* use 3-space indentation as the rest of the codebase
* please pick a shorter function name
* do not use camelCase for the function name, please use all lowercase (this is not 100% consistent throughout the codebase, but prefer names_with_underscores).
* do not use a leading underscore in the function parameter

Also, regarding the behavior: the `include` path _includes_ entries to the list always uses the current directory, right? So for this option, I believe it should probably continue doing the same (i.e. not adding the change in "Add files in CWD only if include option is not set").

Please consider making these changes and I'll be happy to merge this and push a new release!

wesen1 commented:
Hi, sorry for the late response.
I fixed the code style now.

I checked the behavior: If you add one or more explicit `include` paths the `include everything recursively in "."` rule no longer applies.

`runner.file_included`  uses `match_any(runner.configuration.include, filename, true)`.
`match_any()` checks all include patterns for the filename or returns the last parameter (`true`) if the list of include patterns is empty.


While testing my changes I noticed that the `include` paths can not be used to find the directories whose untested files should be included because the values of `include` may be Lua patterns and I don't think there is a way to find all paths that match a Lua pattern, especially since `..` is allowed too.
So instead I "extended" the includeuntestedfiles config value:
* If it is `true` the current behavior is used (Include all untested files in ".")
* Additionally it can be a table with directory and file paths that should be included even if untested


I also added some new tests:

1. dirfilter: To proof that the behavior of auto including works as described above
2. includeuntestedfiles: To check that the option works as expected

The tests revealed some bugs on Windows regarding the file paths of the untested files:
Empty file coverage data is created for every file and the paths of the tested files did sometimes not match the ones of the corresponding empty coverage data.
Fixed that by replacing directory separators by OS specific directory separators at some places.

hishamhm commented:
@wesen1 I fixed the small conflict and merged this PR manually! Thank you!
